sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -997,9 +997,11 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product with Tre
         ("deserialized" -> s.deserialized) ~ ("replication" -> s.replication)
     case n: TreeNode[_] => n.jsonValue
     case o: Option[_] => o.map(parseToJson)
-    // Recursive scan Seq[TreeNode], Seq[Partitioning], Seq[DataType]
+    // Recursive scan Seq[TreeNode], Seq[Partitioning], Seq[DataType], Seq[Product]
     case t: Seq[_] if t.forall(_.isInstanceOf[TreeNode[_]]) ||
-      t.forall(_.isInstanceOf[Partitioning]) || t.forall(_.isInstanceOf[DataType]) =>
+      t.forall(_.isInstanceOf[Partitioning]) ||
+      t.forall(_.isInstanceOf[DataType]) ||
+      t.forall(_.isInstanceOf[Product]) =>
```

Comment:
The `TreeNode` inherits `Product`, I'm wondering if only one check should be kept.

Maybe this change is redundant, the following one which changes `case p if p.productIterator.exists(_.isInstanceOf[TreeNode[_]]) => true` is the key one.

Reply:
Nice catch, I think we can remove the TreeNode check. The code change is for Seq,  without this Seq[Product] will be written as null.